src/combinatorics/hall/basic.lean

```diff
@@ -0,0 +1,230 @@
+set of distinct representatives if and only if every union of `k` of the
+sets has at least `k` elements.
+
+Recall that `s.bUnion t` is the union of all the sets `t i` for `i ∈ s`.
+
+This theorem is bootstrapped from `finset.all_card_le_bUnion_card_iff_exists_injective'`,
+which has the additional constraint that `ι` is a `fintype`.
+-/
+theorem finset.all_card_le_bUnion_card_iff_exists_injective
+  {ι α : Type*} [decidable_eq α] (t : ι → finset α) :
+  (∀ (s : finset ι), s.card ≤ (s.bUnion t).card) ↔
+    (∃ (f : ι → α), function.injective f ∧ ∀ x, f x ∈ t x) :=
+begin
+  split,
+  { intro h,
+    haveI : ∀ (ι' : (finset ι)ᵒᵖ), nonempty ((hall_matchings_functor t).obj ι') :=
+      λ ι', hall_matchings_on.nonempty t h ι'.unop,
+    classical,
+    haveI : Π (ι' : (finset ι)ᵒᵖ), fintype ((hall_matchings_functor t).obj ι') := begin
+      intro ι',
+      rw [hall_matchings_functor],
+      apply_instance,
+    end,
+    obtain ⟨u, hu⟩ := nonempty_sections_of_fintype_inverse_system (hall_matchings_functor t),
+    refine ⟨_, _, _⟩,
+    { exact λ i, (u (opposite.op ({i} : finset ι))).val
+                 ⟨i, by simp only [opposite.unop_op, mem_singleton]⟩, },
+    { intros i i',
+      have subi : ({i} : finset ι) ⊆ {i,i'} := by simp,
+      have subi' : ({i'} : finset ι) ⊆ {i,i'} := by simp,
+      have le : ∀ {s t : finset ι}, s ⊆ t → s ≤ t := λ _ _ h, h,
+      rw [←hu (category_theory.hom_of_le (le subi)).op,
+          ←hu (category_theory.hom_of_le (le subi')).op],
+      let uii' := u (opposite.op ({i,i'} : finset ι)),
+      exact λ h, subtype.mk_eq_mk.mp (uii'.property.1 h), },
+    { intro i,
+      apply (u (opposite.op ({i} : finset ι))).property.2, }, },
+  { rintro ⟨f, hf₁, hf₂⟩ s,
+    rw ←finset.card_image_of_injective s hf₁,
+    apply finset.card_le_of_subset,
+    intro _,
+    rw [finset.mem_image, finset.mem_bUnion],
+    rintros ⟨x, hx, rfl⟩,
+    exact ⟨x, hx, hf₂ x⟩, },
+end
+
+/-- Given a relation such that the image of every singleton set is finite, then the image of every
+finite set is finite. -/
+instance {α β : Type*} [decidable_eq β]
+  (r : α → β → Prop) [∀ (a : α), fintype (rel.image r {a})]
+  (A : finset α) : fintype (rel.image r A) :=
+begin
+  have h : rel.image r A = (A.bUnion (λ a, (rel.image r {a}).to_finset) : set β),
+  { ext, simp [rel.image], },
+  rw [h],
+  apply finset_coe.fintype,
+end
+
+/--
+This is a version of **Hall's Marriage Theorem** in terms of a relation
+between types `α` and `β` such that `α` is finite and the image of
+each `x : α` is finite (it suffices for `β` to be finite; see
+`fintype.all_card_le_filter_rel_iff_exists_injective`).  There is
+a transversal of the relation (an injective function `α → β` whose graph is
+a subrelation of the relation) iff every subset of
+`k` terms of `α` is related to at least `k` terms of `β`.
+
+If `[fintype β]`, then `[∀ (a : α), fintype (rel.image r {a})]` is automatically implied.
```

Comment:
```suggestion
If `[fintype β]`, then `[∀ (a : α), fintype (rel.image r {a})]` is automatically inferred.
```
unless typeclass inference doesn't do it? Would be nice if you could explicit that.

Reply:
There's a point of view that typeclasses are predicates, instances represent implications (well, Horn clauses), and typeclass inference is the mechanism implementing the search.

I'll just change it, since the point of this note could be clearer.